src/analysis/ml/ml.py

```diff
@@ -0,0 +1,137 @@
+            continue
+        
+    # the dataframes per algorithm
+    edge_dataframes = []
+    
+    # the data frame is set up per algorithm and a 1 is set for the edge pair that exists in the algorithm
+    for tuple in edge_tuples:
+
+        dataframe = pd.DataFrame(
+            {
+                str(tuple[0]): 1,
+            }, index= tuple[1]
+        )
+        edge_dataframes.append(dataframe)
+
+    # concating all the data frames together (0 is set for all the edge pairs that don't exist per algorithm)
+    concated_df = pd.concat(edge_dataframes, axis= 1, join = 'outer')
+    concated_df = concated_df.fillna(0) 
+    concated_df = concated_df.astype('int64')
+   
+    return concated_df
+    
+def pca(dataframe: pd.DataFrame, output_png: str, output_file: str):
+
+    df = dataframe.reset_index(drop=True)
+    df = df.transpose() #based on the algortihms rather than the edges 
+    X = df.values
+
+    scaler = StandardScaler()
+    scaler.fit(X) # calc mean and standard deviation 
+    X_scaled = scaler.transform(X)
+
+    # chosing the PCA
+    pca_2 = PCA(n_components=2)
+    pca_2.fit(X_scaled)
+    X_pca_2 = pca_2.transform(X_scaled)
+    variance = pca_2.explained_variance_ratio_ *100
+
+    # making the plot
+    plt.figure(figsize = (10,7))
+    # sns.set(font_scale = 1.5)
+    sns.scatterplot(x = X_pca_2[:,0], y = X_pca_2[:,1], s = 70)
+    plt.title("PCA")
+    plt.xlabel(f"PC1 ({variance[0]:.1f}% variance)")
+    plt.ylabel(f"PC2 ({variance[1]:.1f}% variance)")
+    
+    # saving the pca plot
+    plt.savefig(output_png)
+
+    # saving the principal components
+    with open(output_file, "w") as f: 
+        f.write(str(variance))
+   
+# site sckit learn 
+def plot_dendrogram(model, **kwargs):
+    # Create linkage matrix and then plot the dendrogram
+
+    # create the counts of samples under each node
+    counts = np.zeros(model.children_.shape[0])
+    n_samples = len(model.labels_)
+    for i, merge in enumerate(model.children_):
+        current_count = 0
+        for child_idx in merge:
+            if child_idx < n_samples:
+                current_count += 1  # leaf node
+            else:
+                current_count += counts[child_idx - n_samples]
+        counts[i] = current_count
+
+    linkage_matrix = np.column_stack(
+        [model.children_, model.distances_, counts]
+    ).astype(float)
+
+    # Plot the corresponding dendrogram
+    dendrogram(linkage_matrix, **kwargs)
+
+def hac(dataframe: pd.DataFrame, output_png: str):
+
+    X = dataframe.reset_index(drop=True)
+    X = X.transpose()
+    model = AgglomerativeClustering(linkage='complete', distance_threshold=0, n_clusters=None)
+    model = model.fit(X)
+
+    plt.figure(figsize=(10,7))
+    plt.title("Hierarchical Agglomerative Clustering Dendrogram")
+    algo_names = list(dataframe.columns)
+    plot_dendrogram(model, truncate_mode="level", p=3, labels=algo_names, leaf_rotation=90, leaf_font_size=10)
```

Comment:
```suggestion
    plot_dendrogram(model, truncate_mode=None, labels=algo_names, leaf_rotation=90, leaf_font_size=10, color_threshold=0)
```
Reading carefully through all of the [docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.dendrogram.html) was helpful. Previously we were setting the plot to truncate parts of the dendrogram and group similar graphs. These new settings will show all graphs. I also changed it so that the coloring won't be set automatically because that wasn't meaningful. Everything will be blue.

We'll still need to carefully test that this works on our small graphs. With the real graphs the dendrogram looks strange because some of the graphs are empty.

I also tried avoiding the sklearn code that creates the count data structure but finally understand what that example code is doing. At first, I could tell from the scipy documentation what a "linkage matrix" is. It turns out we actually already get most of it from `AgglomerativeClustering`. ChatGPT was able to explain:
![scipy_linkage](https://user-images.githubusercontent.com/3532898/222873561-113d7543-b598-4117-b1de-98a6bc3573f7.png)

We already have 1, 2, and 3 and only need to construct 4 (`counts`) before passing the matrix to `dendrogram` to plot. That explanation seems to match this part of the scipy docs:
![image](https://user-images.githubusercontent.com/3532898/222873787-3d5c7fa4-f470-4f62-a6cd-fb09d6c9322b.png)